node/src/protocol/cryptography.rs

```diff
@@ -295,9 +297,10 @@ impl CryptographicProtocol for RunningState {
             tracing::warn!(?err, participants = ?self.participants, "running(pre): failed to send encrypted message");
         }
 
+        let participants = ctx.connection_pool().ping().await;
         let mut triple_manager = self.triple_manager.write().await;
         if triple_manager.my_len() < 2 && triple_manager.potential_len() < 10 {
-            triple_manager.generate()?;
+            triple_manager.generate(&participants)?;
```

Comment:
What will happen with triples that were generated when one of the participants was offline?
Should we use the connection pool only to kick members? If we generate triples for n-k participants while k  is still in the participant set, we can complicate the logic significantly.

Reply:
so, while the triple is getting generated here, we will have a list of active participants that were used to generate that triple. What hasn't been accounted for is when the node hops online again and doesn't have these triples nor do we check at all for the threshold. We will need some logic later where we account for these things such that the presignature will only use the participants from the triple generation participants. The big question here is whether that is secure or not

Reply:
if triples were generated when one of the participants was offline:
1) if the participant was included in the set of participants to generate the triple but the participant went offline while triple protocol in progress, then the protocol will be stuck forever, and that triple never gets generated;
2) if the participant was already offline and came back up with a triple that's generated by other nodes. In this case, we need to make sure whenever we generate presignatures, we are only involving the set of participants that are involved in generating both of the 2 triples used for presignature. Similarly when we generate signature, we need to check the participants for the presignatures being used. But here we might have an issue: if the overlap participants between the 2 triples are already <= threshold, should we still proceed? I guess answer is no. We should perhaps set some limit to how big the participants set need to be to start a protocol for triples 🤔 like suppose we are k out of n signing, and we want to always have at least k+x in the participant set for triples, where 2*(k+x) > n + k, i.e. x > (n-k)/2 so there's always an overlap of > k participants in the presignature stage. We want to make sure we still have at least k participants in the final signature stage, otherwise we should choose another pair of triples. 🤔 

Looking at the complication of 2), it might be better

Reply:
I'd say we punt off adding any sort of kicking logic at all. If it becomes a huge problem, each of node can manually call vote_leave on a particular node. It's not the most important thing to consider right now for the beta. Kicking logic can be disastrous with the initial set of nodes we get for genesis where kicking the wrong node can potentially create a state where we have too few nodes to keep on doing resharing.

We should however be cleaning up the node state (triples, presigs) on resharing though.